confgenerator/fluentbit/processors.go

```diff
@@ -21,6 +21,35 @@ import (
 	"strings"
 )
 
+// ParseMultilineComponent constitues the mulltiline_parser components.
+func ParseMultilineComponent(tag string, uid string, field string, languageRules []string) []Component {
+	var components []Component
+	multilineParserName := fmt.Sprintf("multiline.%s.%s", tag, uid)
+	rules := [][2]string{}
+	for _, rule := range languageRules {
+		rules = append(rules, [2]string{"rule", rule})
+	}
+
+	multilineParser := Component{
+		Kind: "MULTILINE_PARSER",
+		Config: map[string]string{
+			"name":          multilineParserName,
+			"type":          "regex",
+			"flush_timeout": "1000",
+		},
+		OrderedConfig: rules,
+	}
+	components = append(components, multilineParser)
+	if field != "" {
+		multilineParser.Config["key_Content"] = field
+		jsonParser, jsonParserName := ParserComponentBase("", "", nil, tag, uid)
+		multilineParser.Config["parser"] = jsonParserName
```

Comment:
pre-parser property is feature provided within the MultilineParser, it doesn't have much to do with the receiver side. Thus it's forward compatible. :)


But you caught something we haven't thought of: if the incoming messages are structured logs already, we shall not use parser to parse it again. Now the behavior is it will, not sure if the json parser is idempotent or not. will need to check.

Reply:
Ugh.. nvm, i think we are good.

structured logs coming from tailing a file is still one line of strings thus it needs to be parsed. There is no structured logs already, as all logs initially come in as strings. A json parser will parse it to be structured within fluentbit.

And to your question again, and no matter what receivers, if the to-be-concatenated logs are in structured json format, the processor needs to know the field.

For example think this as a log file that our agent is tailing, and the to-be-concatenated field is `my_message`. The processor required for it the happen is a parser first parse this json format, then multiline parser that keys on `my_message` to do the concatenation.
```
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "Dec 14 06:41:08 Exception in thread \"main\" java.lang.RuntimeException: Something has gone wrong, aborting!"
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "    at com.myproject.module.MyProject.badMethod(MyProject.java:22)"
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "    at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)"
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "    at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "    at com.myproject.module.MyProject.someMethod(MyProject.java:10)"
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z", "my_message": "    at com.myproject.module.MyProject.main(MyProject.java:6)"
```

Reply:
Will users be able to distinguish between the following 2 needs by specifying the config:
* Structured logs already. Just use `field` to find that field that needs parsing
* Not yet structured logs. Please parse it using JSON then use `field` to find that field that needs parsing

I wonder if we need another config. Or the alternative is to try parse json if it's detected as JSON format (That could introduce additional complexity though)

Reply:
Sync'ed offline, we will punt on having "field"